File: cs50/pset1/credit/credit.c
#include <stdio.h>
#include <cs50.h>

long power(int n);
int digits(long number);

int main(void)
{
    
    // ### 1. Get number input
    long number;
    do
    {
        number = get_long("Number: ");
    }
    while (number < 0);
    
    // ### 2. Check number algorithm
    int sum = 0;
    int n = digits(number);
    
    // Go through all digits
    for (int i = 0; i < n; i++)
    {
        int digit = (number % power(i + 1)) / power(i);
        
        // Multiply every other digit by two and sum
        if (i % 2 == 1)
        {
            int number_new = digit * 2;
            for (int j = 0; j < digits(number_new); j++)
            {
                int digit_new = (number_new % power(j + 1)) / power(j);
                sum += digit_new;
            }
        }
        
        // Sum every other digit
        else
        {
            sum += digit;
        }
    }
    
    // Validate if last digit of sum is 0
    if (sum % 10 != 0)
    {
        printf("INVALID\n");
    }
    
    // ### 3. Check card type
    else
    {
        if ((n == 15) & (number / power(digits(number) - 2) == 34 || number / power(digits(number) - 2) == 37))
        {
            printf("AMEX\n");
        }
        else if ((n == 16) & (number / power(digits(number) - 2) > 50) & (number / power(digits(number) - 2) <= 55))
        {
            printf("MASTERCARD\n");
        }
        else if (((n == 13) || (n == 16)) & (number / power(digits(number) - 1) == 4))
        {
            printf("VISA\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }

}

// Function to get the power (10^x) of a number
long power(int n)
{
    long value = 1;
    for (int i = 0; i < n; i++)
    {
        value = value * 10;
    }
    return value;
}

// Function to get the number of digits of a number
int digits(long number)
{
    int digit = 0;
    while (number > 0)
    {
        number = number / 10;
        digit++;
    }
    return digit;
}

File: cs50/pset6/credit/credit.py
# Packages
from cs50 import get_int



def main():
    # 1. Ask for number and put to string
    number = get_int('Number: ')
    if int(number) < 0:
        number = get_int('Number: ')
    number = str(number)

    # 2. Check number algorithm
    product = 0
    sum = 0
    
    #Compute products
    for i in range(len(number)):
        if i % 2 == 1:
            digit = str(2 * int(number[len(number) - i - 1]))
            for j in range(len(digit)):
                product += int(digit[j])
    
    #Compute sum
        if i % 2 == 0:
            sum += int(number[len(number) - i - 1])
    
    #Test general validity
    test = str(product + sum)
    if int(test[len(test) - 1]) != 0:
        print('INVALID')
    
    # 3. Detect card type
    else:
        if len(number) == 15 and int(number[0:2]) in [34, 37]:
            print('AMEX')
            
        elif len(number) == 16 and int(number[0:2]) in [51, 52, 53, 54, 55]:
            print('MASTERCARD')
    
        elif (len(number) == 13 or len(number) == 16) and int(number[0]) == 4:
            print('VISA')
            
        else:
            print('INVALID')

main()
